operators/pkg/instance-controller/container_logic.go

```diff
@@ -0,0 +1,175 @@
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: pointer.Int32Ptr(1),
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: containers,
+				},
+			},
+		},
+	}, nil
+}
+
+func (r *LabInstanceReconciler) CreateContainerEnvironment(
+	labInstance *crownlabsv1alpha2.Instance,
+	environment *crownlabsv1alpha2.Environment,
+	namespace string,
+	name string,
+	vmStart time.Time) error {
+	// this is added so that all resources created for this Instance are destroyed when the Instance is deleted
+	b := true
+	globalOwnerReference := []metav1.OwnerReference{
+		{
+			APIVersion:         labInstance.APIVersion,
+			Kind:               labInstance.Kind,
+			Name:               labInstance.Name,
+			UID:                labInstance.UID,
+			BlockOwnerDeletion: &b,
+		},
+	}
+	ctx := context.TODO()
+
+	// create Service to expose the vm
+	service := instance_creation.ForgeService(name, namespace, globalOwnerReference)
+
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &service, func() error {
+		return ctrl.SetControllerReference(labInstance, &service, r.Scheme)
+	}); err != nil {
+		setLabInstanceStatus(r, ctx, "Could not create service "+service.Name+" in namespace "+service.Namespace, "Warning", "ServiceNotCreated", labInstance, "", "")
+		return err
+	} else {
+		setLabInstanceStatus(r, ctx, "Service "+service.Name+" correctly created in namespace "+service.Namespace, "Normal", "ServiceCreated", labInstance, "", "")
+	}
+
+	urlUUID := uuid.New().String()
+
+	// create Ingress to manage the service
+	ingress := instance_creation.ForgeIngress(name, namespace, service, urlUUID, r.WebsiteBaseUrl, globalOwnerReference)
+
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &ingress, func() error {
+		return ctrl.SetControllerReference(labInstance, &ingress, r.Scheme)
+	}); err != nil {
+		setLabInstanceStatus(r, ctx, "Could not create ingress "+ingress.Name+" in namespace "+ingress.Namespace, "Warning", "IngressNotCreated", labInstance, "", "")
+		return err
+	} else {
+		setLabInstanceStatus(r, ctx, "Ingress "+ingress.Name+" correctly created in namespace "+ingress.Namespace, "Normal", "IngressCreated", labInstance, "", "")
+	}
+
+	if err := r.createOAUTHlogic(name, labInstance, namespace, globalOwnerReference, urlUUID); err != nil {
+		return err
+	}
+
+	pod, err := CreateGUIContainerPod(name, namespace, environment, labInstance.Name, globalOwnerReference, false)
+
+	if err != nil {
+		return err
+	}
+
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &pod, func() error {
+		return ctrl.SetControllerReference(labInstance, &pod, r.Scheme)
+	}); err != nil {
+		setLabInstanceStatus(r, ctx, "Could not create GUIContainerPod "+pod.Name+" in namespace "+pod.Namespace, "Warning", "GUIContainerPodNotCreated", labInstance, "", "")
+		return err
+	} else {
+		setLabInstanceStatus(r, ctx, "GUIContainerPod "+pod.Name+" correctly created in namespace "+pod.Namespace, "Normal", "GUIContainerPodCreated", labInstance, "", "")
+	}
```

Comment:
This is not the correct usage of the CreateOrUpdate function, as it works in case the resource is created but not if it is updated.

I would suggest to have a look at [this blog post](https://engineering.pivotal.io/post/gp4k-kubebuilder-lessons/), which explains why it is so and what is the correct pattern that should be used. Additionally, you can also refer to the [official documentation](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate) and to other examples in the [tenant controller](https://github.com/netgroup-polito/CrownLabs/blob/master/operators/pkg/tenant-controller/tenant_controller.go).